server/journeys/referral/details/schemas.ts

```diff
@@ -57,23 +57,33 @@ export const schemaFactory = (csipApiService: CsipApiService) => async (req: Req
           .max(new Date(), { message: INCIDENT_DATE_FUTURE_MSG }),
       )
       .transform(dateString => dateString.toISOString().substring(0, 10)),
-    hour: z.string().transform(val => (val?.length ? parse24Hour(val) : null)),
-    minute: z.string().transform(val => (val?.length ? parseMinute(val) : null)),
+    hour: z.string(),
+    minute: z.string(),
     incidentLocation: z
       .string({ message: INCIDENT_LOCATION_MSG })
       .transform(validateAndTransformReferenceData(incidentLocationMap, INCIDENT_LOCATION_MSG)),
     incidentType: z
       .string({ message: INCIDENT_TYPE_MSG })
       .transform(validateAndTransformReferenceData(incidentTypeMap, INCIDENT_TYPE_MSG)),
   })
-    .refine(val => (!val.hour && !val.minute) || (val.hour?.success && val.minute?.success), {
-      message: 'Enter a time using the 24-hour clock',
-      path: ['incidentTime'],
+    .superRefine((val, ctx) => {
+      if (
+        !!val.hour?.length !== !!val.minute?.length ||
+        (val.hour?.length && val.minute?.length && (!parse24Hour(val.hour).success || !parseMinute(val.minute).success))
+      ) {
+        ctx.addIssue({
+          code: z.ZodIssueCode.custom,
+          message: 'Enter a time using the 24-hour clock',
+          path: ['incidentTime'],
+        })
+      }
+    })
+    .transform(val => {
+      const hour = val.hour?.length ? parse24Hour(val.hour) : null
```

Comment:
Can we do the checks and transformation in `transform`, calling `addIssue` there if needed, instead of having superRefine then transform and doing very similar operations in both?

Reply:
the helper function does what it says on the tin. it `always refine`. but unfortunately does not `always transform`.